sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala

```diff
@@ -174,6 +174,20 @@ object DecimalPrecision extends TypeCoercionRule {
       CheckOverflow(Pmod(promotePrecision(e1, widerType), promotePrecision(e2, widerType)),
         resultType, nullOnOverflow)
 
+    case expr @ IntegralDivide(
+        e1 @ DecimalType.Expression(p1, s1), e2 @ DecimalType.Expression(p2, s2)) =>
+      val widerType = widerDecimalType(p1, s1, p2, s2)
+      val promotedExpr =
+        IntegralDivide(promotePrecision(e1, widerType), promotePrecision(e2, widerType))
+      if (expr.dataType.isInstanceOf[DecimalType]) {
+        // This follows division rule
+        val intDig = p1 - s1 + s2
+        // No precision loss can happen as the result scale is 0, so only overflow can happen
+        CheckOverflow(promotedExpr, DecimalType.bounded(intDig, 0), nullOnOverflow)
```

Comment:
Any test for the overflow case?

Reply:
it is very hard to cause an overflow, since the scale is 0, so the overflow can happen only in the `promotePrecision` phase, in those casts... maybe we can get rid of this `CheckOverflow`, although I think for safety it is good to have it. WDYT?

Reply:
For example `select div(-9223372036854775808, -1)` should cause overflow with `spark.sql.legacy.integralDivide.returnBigint=true`? But, in case of `spark.sql.legacy.integralDivide.returnBigint=false`, IIUC we have no chance to hit overflow. Actually, I think we should promote it without overflow.

Reply:
when `spark.sql.legacy.integralDivide.returnBigint=true` anyway, the overflow should be checked by the `toLong` method, ie. the "cast" to long, which is yet to be handled in a different PR.

Let me remove this `CheckOverflow` then, thanks.

Reply:
> when spark.sql.legacy.integralDivide.returnBigint=true anyway, the overflow should be checked by the toLong method, ie. the "cast" to long, which is yet to be handled in a different PR.

Yea, it looks ok to me. Thanks for the check!

Reply:
> Eg, if the operands are decimal(35, 0) and decimal(5, 3), they are promoted to decimal(35, 3), but the correct return type should be decimal(38,0)

Ah, I see. Thanks for the explanation! I missed that case. But, even if you know that condition, you cannot write a test case for that?

Reply:
> Yea, it looks ok to me. Thanks for the check!

thank you for the comment and checking this out.

> even if you know that condition, you cannot write a test case for that?

I can, the point is that we are not testing any overflow. But I can add a test like the one mentioned above and I am doing it, thanks.

Reply:
Many Thanks!!

Reply:
Thank you!